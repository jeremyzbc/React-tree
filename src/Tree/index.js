import React, { Component } from 'react';
import { ExpandLess, ExpandMore } from '@material-ui/icons';
import PropTypes from 'prop-types';

export default class Tree extends Component {
  constructor(props) {
    super(props);
    this.state = {
      tree: props.tree
    };
  }

  static getDerivedStateFromProps(nextProps) {
    return {
      tree: nextProps.source
    };
  }

  updateTreeByBranchId = (tree, id, callback) => {
    this.find = false;
    return tree.map(branch => {
      if (this.find) {
        return branch;
      } else if (branch.id == id) {
        this.find = true;
        return callback(branch);
      } else if (branch.children !== undefined && branch.children.length > 0) {
        branch.children = this.updateTreeByBranchId(
          branch.children,
          id,
          callback
        );
        return branch;
      }
      return branch;
    });
  };

  expandToggle = id => {
    const callback = branch => {
      branch.isExpand = !branch.isExpand;
      return branch;
    };
    const tree = this.updateTreeByBranchId(this.state.tree, id, callback);
    this.setState({ tree });
  };

  selectToggle = id => {
    const callback = branch => {
      branch.isSelect = !branch.isSelect;

      // select/unselect parent branch will select/unselect all children branch
      if (branch.children !== undefined && branch.children.length > 0) {
        branch.children = this.updateChildrenFromParent(branch);
      }
      return branch;
    };
    let tree = this.updateTreeByBranchId(this.state.tree, id, callback);

    // select/unselect child branch may affect select/unselect parent branch
    tree = this.updateParentFromChildren(tree);

    // fire event for parent Component once tree selection is updated
    this.setState({ tree }, () => {
      if (typeof this.props.onTreeChange === 'function')
        this.props.onTreeChange(tree);
    });
  };

  updateChildrenFromParent = parentBranch =>
    parentBranch.children.map(childBranch => {
      childBranch.isSelect = parentBranch.isSelect;
      if (
        childBranch.children !== undefined &&
        childBranch.children.length > 0
      ) {
        childBranch.children = this.updateChildrenFromParent(childBranch);
      }
      return childBranch;
    });

  updateParentFromChildren = tree => {
    // select one of child branches will select parent branch, unselect all child branches will unselect parent branch
    return tree.map(branch => {
      if (branch.children !== undefined && branch.children.length > 0) {
        branch.children = this.updateParentFromChildren(branch.children);

        if (branch.children.every(childBranch => !childBranch.isSelect)) {
          branch.isSelect = false;
        } else {
          branch.isSelect = true;
        }
      }
      return branch;
    });
  };

  makeTree = tree => (
    <div style={{ marginLeft: '1rem' }}>
      {tree.map(branch => (
        <div key={branch.id}>
          {this.props.selectable ? (
            <input
              type="checkbox"
              checked={branch.isSelect}
              onChange={() => this.selectToggle(branch.id)}
            />
          ) : null}
          <span>{branch.label}</span>
          {branch.children && branch.children.length > 0 ? (
            branch.isExpand ? (
              <>
                <ExpandLess onClick={() => this.expandToggle(branch.id)} />
                {this.makeTree(branch.children)}
              </>
            ) : (
              <ExpandMore onClick={() => this.expandToggle(branch.id)} />
            )
          ) : null}
        </div>
      ))}
    </div>
  );

  render() {
    return this.makeTree(this.state.tree);
  }
}

Tree.propTypes = {
  source: PropTypes.arrayOf(PropTypes.object).isRequired
};
