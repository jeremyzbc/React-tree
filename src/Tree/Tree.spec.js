import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Tree from './index';
import { default as mockData } from '../example/source';

describe('Tree Component', () => {
  it('Should match snapshot', () => {
    const tree = renderer.create(<Tree source={mockData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should toggle lower level branches when expand icon is clicked', () => {
    const wrapper = shallow(<Tree source={mockData} />);
    const branchNumberBefore = wrapper.find('div').length;
    const expandToggles = wrapper.find('i');
    const randomIndex = Math.floor(Math.random() * expandToggles.length);
    const expandToggle = expandToggles.at(randomIndex);
    expandToggle.simulate('click');
    const branchNumberAfter = wrapper.find('div').length;
    expect(branchNumberAfter).not.toEqual(branchNumberBefore);
  });

  it('Should select child branches when parent branch is selected', () => {
    const wrapper = shallow(<Tree source={mockData} />);
    const getCheckbox = label =>
      wrapper
        .findWhere(n => n.text() === label && n.type() === 'span')
        .closest('div')
        .find('[type="checkbox"]');

    const parentCheckbox = getCheckbox('Australia');
    const childrenCheckboxs = [
      getCheckbox('NSW'),
      getCheckbox('QLD'),
      getCheckbox('Victoria'),
      getCheckbox('Western Australia'),
      getCheckbox('ACT')
    ];

    parentCheckbox.simulate('change');

    childrenCheckboxs.every(cb => expect(cb.props().checked).toEqual(true));
  });
});
