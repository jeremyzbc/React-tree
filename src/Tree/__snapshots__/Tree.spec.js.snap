// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Component Should match snapshot 1`] = `
<div
  style={
    Object {
      "padding": ".6rem 1rem",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "1.5rem",
      }
    }
  >
    <input
      onChange={[Function]}
      type="checkbox"
    />
    <span>
      Australia
    </span>
    <i
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-1"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 8l-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </i>
  </div>
  <div
    style={
      Object {
        "padding": ".6rem 1rem",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "height": "1.5rem",
        }
      }
    >
      <input
        onChange={[Function]}
        type="checkbox"
      />
      <span>
        NSW
      </span>
      <i
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-1"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </i>
    </div>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "height": "1.5rem",
        }
      }
    >
      <input
        onChange={[Function]}
        type="checkbox"
      />
      <span>
        QLD
      </span>
      <i
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-1"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </i>
    </div>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "height": "1.5rem",
        }
      }
    >
      <input
        onChange={[Function]}
        type="checkbox"
      />
      <span>
        Victoria
      </span>
      <i
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-1"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </i>
    </div>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "height": "1.5rem",
        }
      }
    >
      <input
        onChange={[Function]}
        type="checkbox"
      />
      <span>
        Western Australia
      </span>
      <i
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-1"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </i>
    </div>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "height": "1.5rem",
        }
      }
    >
      <input
        onChange={[Function]}
        type="checkbox"
      />
      <span>
        ACT
      </span>
    </div>
  </div>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "1.5rem",
      }
    }
  >
    <input
      onChange={[Function]}
      type="checkbox"
    />
    <span>
      New Zealand
    </span>
  </div>
</div>
`;
